File: Chapter 04/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter04/section_4_2_downward_dependency/after/SqlOrderRepository.java
package com.grokkingsoftwarearchitecture.chapter04.section_4_2_downward_dependency.after;

import com.grokkingsoftwarearchitecture.chapter04.shared.LogManager;

/**
 * DATA ACCESS LAYER.
 * Implements the interface.
 */
public class SqlOrderRepository implements OrderRepository {
    @Override
    public void save(Order order) {
        LogManager.info(SqlOrderRepository.class, "(After Refactor) Saving order to SQL...");
    }
}

File: Chapter 04/Python/section_4_2_downward_dependency/after/sql_order_repository.py
from .order_repository import OrderRepository
from .order import Order
from shared.log_manager import LogManager

class SqlOrderRepository(OrderRepository):
    """
    DATA ACCESS LAYER (LOW-LEVEL DETAIL).
    ARCHITECTURE NOTE: This is a concrete implementation. 
    It 'plugs into' the architecture by fulfilling the 
    OrderRepository contract.
    """
    def save(self, order: Order):
        LogManager.info("SqlOrderRepository", "(After Refactor) Saving order to SQL...")